File: Turquoise/NNTPClient/TQNNTPArticleForest.m

#import "TQNNTPArticleForest.h"

#import "NSString+TQEncoding.h"

@implementation TQNNTPArticleForest {
  NSMutableDictionary<NSString *, TQNNTPArticle *> *_messageIds;
  NSMutableArray<TQNNTPArticle *> *_rootArticles;
}

- (instancetype)initWithArticles:(NSArray<TQNNTPArticle *> *)articles {
  if (!articles) {
    return nil;
  }

  self = [super init];
  if (self) {
    _rootArticles = [NSMutableArray array];
    _messageIds = [NSMutableDictionary dictionary];

    for (TQNNTPArticle *article in articles) {
      if (article.messageId) {
        _messageIds[article.messageId] = article;
      }

      if (!article.parentArticle) {
        // the parent article may have been deleted. if so, try to find the next available ancestor.
        for (NSString *referenceMessageId in article.references.reverseObjectEnumerator) {
          TQNNTPArticle *ancestorArticle = _messageIds[referenceMessageId];
          if (ancestorArticle) {
            article.parentArticle = ancestorArticle;
            [ancestorArticle.childArticles addObject:article];
            // TODO: where exactly should we "insert" this new child into the ancestor's children?
            break;
          }
        }
      }
      if (!article.parentArticle) {
        [_rootArticles addObject:article];
      }
    }

    [_rootArticles sortUsingComparator:^NSComparisonResult(TQNNTPArticle *_Nonnull obj1, TQNNTPArticle *_Nonnull obj2) {
      return [@(obj2.articleNo) compare:@(obj1.articleNo)];
    }];
  }

  return self;
}

- (NSUInteger)numArticles {
  return _messageIds.allKeys.count;
}

- (NSArray *)trees {
  return _rootArticles;
}

- (NSArray<TQNNTPArticle *> *)expandedForest {
  NSMutableArray *forest = [NSMutableArray array];
  for (TQNNTPArticle *rootArticle in _rootArticles) {
    [forest addObjectsFromArray:[self expandedForestFromNode:rootArticle]];
  }

  return forest;
}

- (NSMutableArray<TQNNTPArticle *> *)expandedForestFromNode:(TQNNTPArticle *)startingNode {
  if (!startingNode) {
    return nil;
  }

  NSMutableArray *forest = [NSMutableArray arrayWithObject:startingNode];
  for (TQNNTPArticle *child in startingNode.childArticles) {
    [forest addObjectsFromArray:[self expandedForestFromNode:child]];
  }

  return forest;
}

#pragma mark - Debug

- (void)printForest {
  for (TQNNTPArticle *root in _rootArticles) {
    [self printTreeNode:root level:0];
  }
}

- (void)printTreeNode:(TQNNTPArticle *)article level:(NSInteger)level {
  NSMutableString *mutString = [NSMutableString string];
  if (level == 0) {
    [mutString appendString:@"*"];
  } else {
    for (NSInteger i = 0; i < level; i++) {
      [mutString appendString:@"-"];
    }
  }
  [mutString appendFormat:@"%@ (%ld)", [article.subject tq_decodedString] , article.articleNo];

  NSLog(@"%@", mutString);

  for (TQNNTPArticle *child in article.childArticles) {
    [self printTreeNode:child level:(level + 1)];
  }
}

@end

File: Turquoise/TQNNTPArticleForest.swift
import Foundation

@objc class TQNNTPArticleForest : NSObject {
  var trees: [TQNNTPArticle] {
    get {
      return self.rootArticles
    }
  }
  var numArticles: Int {
    get {
      return self.messageIds.count
    }
  }
  private var messageIds: [String : TQNNTPArticle]
  private var rootArticles: [TQNNTPArticle]

  init(articles: [TQNNTPArticle]) {
    self.rootArticles = []
    self.messageIds = [:]

    for article in articles {
      self.messageIds[article.messageId] = article

      if article.parentArticle == nil {
        // the parent article may have been deleted. if so, try to find the next available ancestor.
        for referenceMessageId in article.references.reversed() {
          if let ancestorArticle = self.messageIds[referenceMessageId] {
            article.parentArticle = ancestorArticle
            ancestorArticle.addChildArticle(article)
            // TODO: where exactly should we "insert" this new child into the ancestor's children?
            break
          }
        }
      }
      if article.parentArticle == nil {
        self.rootArticles.append(article)
      }
    }

    self.rootArticles.sort { $1.articleNo < $0.articleNo }
  }

  func expandedForest() -> [TQNNTPArticle] {
    var forest: [TQNNTPArticle] = []
    for rootArticle in self.rootArticles {
      forest.append(contentsOf: self.expandedForest(from: rootArticle))
    }
    return forest
  }

  private func expandedForest(from node: TQNNTPArticle?) -> [TQNNTPArticle] {
    guard let node = node else {
      return []
    }

    var forest = [node]
    for child in node.childArticles {
      forest.append(contentsOf: self.expandedForest(from: child))
    }
    return forest
  }

  // MARK: - Debug
  func printForest() {
    for root in self.rootArticles {
      self.printTreeNode(root, level: 0)
    }
  }

  func printTreeNode(_ article: TQNNTPArticle, level: Int) {
    var string = ""
    if level == 0 {
      string += "*"
    } else {
      string += String(repeatElement("-", count: level))
    }
    string += "\(article.subject.tq_decodedString) (\(article.articleNo)"
    printDebug(string)

    for child in article.childArticles {
      self.printTreeNode(child, level: (level + 1))
    }
  }
}
